server/src/main/java/com/tencent/rss/server/buffer/ShuffleBuffer.java

```diff
@@ -0,0 +1,285 @@
+        isValid,
+        this);
+    inFlushBlockMap.put(eventId, spBlocks);
+    blocks.clear();
+    size = 0;
+    return event;
+  }
+
+  public List<ShufflePartitionedBlock> getBlocks() {
+    return blocks;
+  }
+
+  public long getSize() {
+    return size;
+  }
+
+  public boolean isFull() {
+    return size > capacity;
+  }
+
+  public synchronized void clearInFlushBuffer(long eventId) {
+    inFlushBlockMap.remove(eventId);
+  }
+
+  @VisibleForTesting
+  public Map<Long, List<ShufflePartitionedBlock>> getInFlushBlockMap() {
+    return inFlushBlockMap;
+  }
+
+  // 1. generate buffer segments and other info: if blockId exist, start with which eventId
+  // 2. according to info from step 1, generate data
+  // todo: if block was flushed, it's possible to get duplicated data
+  public synchronized ShuffleDataResult getShuffleData(
+      long lastBlockId, int readBufferSize) {
+    try {
+      List<BufferSegment> bufferSegments = Lists.newArrayList();
+      List<ShufflePartitionedBlock> readBlocks = Lists.newArrayList();
+      updateBufferSegmentsAndResultBlocks(
+          lastBlockId, readBufferSize, bufferSegments, readBlocks);
+      if (!bufferSegments.isEmpty()) {
+        BufferSegment bufferSegment = bufferSegments.get(bufferSegments.size() - 1);
+        int length = bufferSegment.getOffset() + bufferSegment.getLength();
+        byte[] data = new byte[length];
+        // copy result data
+        updateShuffleData(readBlocks, data);
+        return new ShuffleDataResult(data, bufferSegments);
+      }
+    } catch (Exception e) {
+      LOG.error("Exception happened when getShuffleData in buffer", e);
+    }
+    return new ShuffleDataResult();
+  }
+
+  // here is the rule to read data in memory:
+  // 1. read from inFlushBlockMap order by eventId asc, then from blocks
+  // 2. if can't find lastBlockId, means related data may be flushed to storage, repeat step 1
+  private void updateBufferSegmentsAndResultBlocks(
+      long lastBlockId,
+      long readBufferSize,
+      List<BufferSegment> bufferSegments,
+      List<ShufflePartitionedBlock> resultBlocks) {
+    long nextBlockId = lastBlockId;
+    List<Long> sortedEventId = sortFlushingEventId();
+    int offset = 0;
+    boolean hasLastBlockId = false;
+    // read from inFlushBlockMap first to make sure the order of
+    // data read is according to the order of data received
+    if (!inFlushBlockMap.isEmpty()) {
+      for (Long eventId : sortedEventId) {
+        // update bufferSegments with current blocks in flush map
+        CachedBlocksReadInfo cachedBlocksReadInfo = updateBufferSegmentsWithCachedBlocks(
+            offset, inFlushBlockMap.get(eventId), nextBlockId, readBufferSize,
+            bufferSegments, resultBlocks);
+        offset = cachedBlocksReadInfo.getOffsetInResultData();
+        // if has data or find last blockId, read from begin with next cached blocks
+        if (offset > 0 || cachedBlocksReadInfo.isHasLastBlockId()) {
```

Comment:
We judge hasLastBlockId is enough?

Reply:
offset > 0 is deal with lastBlockId = -1, it can be simplifier to keep hasLastBlockId only.